tensorflow_quantum/core/ops/tfq_simulate_expectation_op.cc

```diff
@@ -129,26 +164,86 @@ class TfqSimulateExpectationOp : public tensorflow::OpKernel {
                                               scratch, &exp_v));
-        output_tensor(i, j) = exp_v;
+        (*output_tensor)(i, j) = exp_v;
       }
     }
-    // just to be on the safe side.
     sv.release();
     scratch.release();
-    qsim_circuits.clear();
-    fused_circuits.clear();
-    num_qubits.clear();
-    maps.clear();
-    pauli_sums.clear();
-    programs.clear();
+  }
+
+  void ComputeSmall(
+      const std::vector<int>& num_qubits, const int max_num_qubits,
+      const std::vector<std::vector<qsim::GateFused<QsimGate>>>& fused_circuits,
+      const std::vector<std::vector<PauliSum>>& pauli_sums,
+      tensorflow::OpKernelContext* context,
+      tensorflow::TTypes<float, 1>::Matrix* output_tensor) {
+    const auto tfq_for = qsim::SequentialFor(1);
+    using Simulator = qsim::Simulator<const qsim::SequentialFor&>;
+    using StateSpace = Simulator::StateSpace;
+    using State = StateSpace::State;
+
+    const int output_dim_op_size = output_tensor->dimension(1);
+
+    auto DoWork = [&](int start, int end) {
+      int old_batch_index = -2;
+      int cur_batch_index = -1;
+      int largest_nq = 1;
+      int cur_op_index;
+
+      State sv = StateSpace(largest_nq, tfq_for).CreateState();
+      State scratch = StateSpace(largest_nq, tfq_for).CreateState();
+      for (int i = start; i < end; i++) {
+        cur_batch_index = i / output_dim_op_size;
+        cur_op_index = i % output_dim_op_size;
+
+        const int nq = num_qubits[cur_batch_index];
+        Simulator sim = Simulator(nq, tfq_for);
+        StateSpace ss = StateSpace(nq, tfq_for);
+
+        // (#679) Just ignore empty program
+        if (fused_circuits[cur_batch_index].size() == 0) {
+          (*output_tensor)(cur_batch_index, cur_op_index) = -2.0;
+          continue;
+        }
+
+        if (cur_batch_index != old_batch_index) {
+          // We've run into a new wavefunction we must compute.
+          // Only compute a new wavefunction when we have to.
+          if (nq >= largest_nq) {
+            sv = ss.CreateState();
+            scratch = ss.CreateState();
+            largest_nq = nq;
+          }
+          // no need to update scratch_state since ComputeExpectation
+          // will take care of things for us.
+          ss.SetStateZero(sv);
+          for (int j = 0; j < fused_circuits[cur_batch_index].size(); j++) {
+            qsim::ApplyFusedGate(sim, fused_circuits[cur_batch_index][j], sv);
+          }
+        }
+
+        float exp_v = 0.0;
+        OP_REQUIRES_OK(context, ComputeExpectationQsim(
+                                    pauli_sums[cur_batch_index][cur_op_index],
+                                    sim, ss, sv, scratch, &exp_v));
+        (*output_tensor)(cur_batch_index, cur_op_index) = exp_v;
+        old_batch_index = cur_batch_index;
+      }
+      sv.release();
+      scratch.release();
+    };
+
+    const int num_cycles = 200 * (1 << max_num_qubits);
```

Comment:
is this intended?

1<<25 = 33554432, and 200 * (1<<25) = 6710886400

Reply:
Yes, one unit of work is much larger in this case because it is simulating a whole circuit, which takes lots of time.